### PR TITLE
HACK week: migration to TypeScript (client/utils/currency) 

### PR DESCRIPTION
**Pull request by mgascam**
Migration to TS of client/card-readers/utils/currency folder component.

#### Testing instructions

* All CI checks should pass

### CODE REVIEW

File: client/utils/currency/index.ts [hunk 1]
```diff
@@ -2,12 +2,19 @@
...
 import { find, trimEnd, endsWith } from 'lodash';
+import {
+	CurrencyNames,
+	ExchangeRateDenomination,
+} from 'wcpay/types/utils/currency';
 
 const currencyData = getCurrencyData();
 
-const currencyNames = {
+const currencyNames: CurrencyNames = {
```
Comment (kalessil) [inline client/utils/currency/index.ts:17]: IMO we should inline the type here: `: Record< string, string >`

File: client/utils/currency/index.ts [hunk 2]
```diff
@@ -2,12 +2,19 @@
...
-import Currency, { getCurrencyData } from '@woocommerce/currency';
+import Currency, {
+	CurrencyInterface,
+	getCurrencyData,
+} from '@woocommerce/currency';
 import { find, trimEnd, endsWith } from 'lodash';
+import {
+	CurrencyNames,
+	ExchangeRateDenomination,
+} from 'wcpay/types/utils/currency';
```
Comment (kalessil) [inline client/utils/currency/index.ts:13]: This import should be under the internal dependencies block.

File: client/utils/currency/index.ts [hunk 3]
```diff
@@ -69,12 +79,34 @@ export const getCurrency = ( currencyCode, baseCurrencyCode = null ) => {
...
 	return wcpaySettings.zeroDecimalCurrencies.includes(
 		currencyCode.toLowerCase()
 	);
 };
 
+function composeFallbackCurrency(
+	amount: number,
+	currencyCode: string,
+	isZeroDecimal: boolean
+) {
```
Comment (kalessil) [inline client/utils/currency/index.ts:92]: We can declare return type here.

File: client/utils/currency/index.ts [hunk 4]
```diff
@@ -117,13 +149,18 @@ export const formatCurrency = (
  *
  * @return {string} formatted currency representation with the currency code suffix
  */
-const appendCurrencyCode = ( formatted, currencyCode ) => {
+const appendCurrencyCode = ( formatted: string, currencyCode: string ) => {
```
Comment (kalessil) [inline client/utils/currency/index.ts:152]: We can declare return type here.

File: client/utils/currency/index.ts [hunk 5]
```diff
@@ -117,13 +149,18 @@ export const formatCurrency = (
...
 	if ( -1 === formatted.toString().indexOf( currencyCode ) ) {
 		formatted = formatted + ' ' + currencyCode;
 	}
 	return formatted;
 };
 
+function removeCurrencySymbol( formatted: string ) {
```
Comment (kalessil) [inline client/utils/currency/index.ts:159]: We can declare return type here.

File: client/utils/currency/index.ts [hunk 6]
```diff
@@ -135,48 +172,31 @@ const appendCurrencyCode = ( formatted, currencyCode ) => {
...
-	return `${ formatExplicitCurrency(
-		fromAmount,
-		from.currency,
-		true
-	) } → ${ formatExchangeRate( from, to ) }: ${ formatExplicitCurrency(
-		Math.abs( to.amount ),
-		to.currency
-	) }`;
-};
+function trimEndingZeroes( formattedCurrencyAmount = '' ) {
```
Comment (kalessil) [inline client/utils/currency/index.ts:187]: We can declare return type here.

File: client/utils/currency/index.ts [hunk 7]
```diff
@@ -135,48 +172,31 @@ const appendCurrencyCode = ( formatted, currencyCode ) => {
...
+			endsWith( chunk, '0' ) ? trimEnd( chunk, '0' ) : chunk
+		)
+		.join( ' ' );
+}
 
-function formatExchangeRate( from, to ) {
+function formatExchangeRate(
+	from: ExchangeRateDenomination,
+	to: ExchangeRateDenomination
+) {
```
Comment (kalessil) [inline client/utils/currency/index.ts:199]: We can declare return type here.

**Review by kalessil — CHANGES_REQUESTED** [review]
Nice! Left some comments around adding more types.

File: client/utils/currency/index.ts (hunk 4 above)
Comment (mgascam) [reply client/utils/currency/index.ts]: Done

File: client/utils/currency/index.ts (hunk 6 above)
Comment (mgascam) [reply client/utils/currency/index.ts]: Done

File: client/utils/currency/index.ts (hunk 7 above)
Comment (mgascam) [reply client/utils/currency/index.ts]: Done

**Review by kalessil — APPROVED** [review]
LGTM!

**Pull request merged**